shared/components/leaders/leaders.component.js

```diff
@@ -0,0 +1,315 @@
+import Panel from './../../lib/base_classes/panel';
+import template from './leaders.rt.html';
+
+class LeadersComponent extends Panel {
+
+  constructor(props, context){
+    super(props, context);
+    let leaders = this;
+    leaders.state = {
+      limit: 20,
+      offset: 0,
+      list: [],
+      cache: [],
+      total_count: 0,
+      trigger_update: true,
+      is_loading: true,
+      all_loaded: false,
+      selected_location: {
+        city: "",
+        state: ""
+      },
+      locations: [],
+      show_locations_list: false
+    }
+  }
+
+  get category_identifier() {
+    let leaders = this,
+    id;
+    switch (leaders.current_route_name){
+      case 'Travel':
+        id = 'result_transport_total';
+        break;
+      case 'Home':
+        id = 'result_housing_total';
+        break;
+      case 'Food':
+        id = 'result_food_total';
+        break;
+      case 'Shopping':
+        id = 'result_shopping_total';
+        break;
+      default:
+        id = 'result_grand_total';
+        break;
+    }
+    return id;
+  }
+
+  get selected_location() {
+    if (this.is_location_filtered) {
+      return this.state.selected_location
+    }
+  }
+
+  get is_location_filtered() {
+    if (Object.values(this.state.selected_location.city).length != 0) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  get list(){
+    return this.state.list;
+  }
+
+  get is_loading(){
+    return this.state.is_loading;
+  }
+
+  get footprint_title(){
+    let leaders = this,
+    title;
+    switch (leaders.current_route_name){
+      case 'Travel':
+        title = leaders.t('leaders.travel_footprint');
+        break;
+      case 'Home':
+        title = leaders.t('leaders.home_footprint');
+        break;
+      case 'Food':
+        title = leaders.t('leaders.food_footprint');
+        break;
+      case 'Shopping':
+        title = leaders.t('leaders.shopping_footprint');
+        break;
+      default:
+        title = leaders.t('leaders.total_footprint');
+        break;
+    }
+    return title;
+  }
+
+  get total_count_reached(){
+    if (this.state.offset === 0 && this.state.offset === 20) {
+      return this.state.limit >= this.state.total_count
+    } else {
+      return this.state.offset + this.state.limit >= this.state.total_count;
+    }
+  }
+
+  get show_locations_list() {
+    return this.state.show_locations_list;
+  }
+
+  componentDidMount() {
+    let leaders = this;
+    if (leaders.state_manager.state.leaders_chart.show) {
+      leaders.retrieveLocations();
+      leaders.retrieveAndShow();
+    }
+    leaders.state_manager.state.leaders_chart.current_route = leaders.current_route_name;
+  }
+
+  componentDidUpdate() {
+    let leaders = this;
+    if (leaders.state_manager.state.leaders_chart.current_route != leaders.current_route_name) {
+      leaders.state_manager.state.leaders_chart.current_route = leaders.current_route_name
```

Comment:
For readability, I like to put these methods in something like `LeadersComponent#get has_route_updated` or something like that, but that's a pretty minor stylistic preference.